src/server/scripts/Northrend/ChamberOfAspects/ObsidianSanctum/boss_sartharion.cpp

```diff
@@ -190,377 +205,415 @@ class boss_sartharion : public CreatureScript
+            dragonsCount = 0;
             volcanoBlows.clear();
-
-            if (pInstance)
-            {
-                pInstance->SetData(BOSS_SARTHARION_EVENT, NOT_STARTED);
-                pInstance->DoRemoveAurasDueToSpellOnPlayers(SPELL_TWILIGHT_SHIFT);
-            }
+            instance->DoRemoveAurasDueToSpellOnPlayers(SPELL_TWILIGHT_SHIFT);
         }
 
-        void DoAction(int32 param)
+        void DoAction(int32 param) override
         {
             if (param == ACTION_DRAKE_DIED)
-                me->CastSpell(me, SPELL_SARTHARION_TWILIGHT_REVENGE, true);
+                DoCastSelf(SPELL_SARTHARION_TWILIGHT_REVENGE, true);
         }
 
-        void EnterCombat(Unit*  /*pWho*/)
+        void EnterCombat(Unit* /*pWho*/) override
         {
-            me->CastSpell(me, SPELL_SARTHARION_PYROBUFFET, true);
-            me->SetInCombatWithZone();
+            _EnterCombat();
+            DoCastSelf(SPELL_SARTHARION_PYROBUFFET, true);
             Talk(SAY_SARTHARION_AGGRO);
-            if (pInstance)
-                pInstance->SetData(BOSS_SARTHARION_EVENT, IN_PROGRESS);
 
+            // Combat events
             events.ScheduleEvent(EVENT_SARTHARION_CAST_CLEAVE, 7000);
             events.ScheduleEvent(EVENT_SARTHARION_CAST_FLAME_BREATH, 15000);
             events.ScheduleEvent(EVENT_SARTHARION_CAST_TAIL_LASH, 11000);
-            events.ScheduleEvent(EVENT_SARTHARION_SUMMON_LAVA, 20000);
-            events.ScheduleEvent(EVENT_SARTHARION_LAVA_STRIKE, 5000);
-            events.ScheduleEvent(EVENT_SARTHARION_HEALTH_CHECK, 10000);
-            events.ScheduleEvent(EVENT_SARTHARION_BERSERK, 900000);
-            events.ScheduleEvent(EVENT_SARTHARION_BOUNDARY, 1000);
 
-            StoreDragons();
+            // Extra events
+            extraEvents.ScheduleEvent(EVENT_SARTHARION_SUMMON_LAVA, 20000);
+            extraEvents.ScheduleEvent(EVENT_SARTHARION_LAVA_STRIKE, 5000);
+            extraEvents.ScheduleEvent(EVENT_SARTHARION_HEALTH_CHECK, 10000);
+            extraEvents.ScheduleEvent(EVENT_SARTHARION_BERSERK, 900000);
+            extraEvents.ScheduleEvent(EVENT_SARTHARION_BOUNDARY, 1000);
+
+            // Store dragons
+            for (uint8 i = 0; i < 3; ++i)
+            {
+                Creature* dragon = ObjectAccessor::GetCreature(*me, instance->GetData64(dragons.at(i)));
+                if (!dragon || !dragon->IsAlive())
+                {
+                    continue;
+                }
+
+                dragon->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_NPC);
+                ++dragonsCount;
+                me->AddLootMode(1 << dragonsCount);
+
+                dragon->SetFullHealth();
+                switch (dragons.at(i))
+                {
+                    case DATA_TENEBRON:
+                    {
+                        dragon->CastSpell(dragon, SPELL_POWER_OF_TENEBRON, true);
+                        extraEvents.ScheduleEvent(EVENT_SARTHARION_CALL_TENEBRON, 10000);
```

Comment:
shall we wrap numbers such as `10000` into constants?

Reply:
Thats not necessary i think